---
title: "rg results summary tables template"
author: "Abigail ter Kuile, Christopher Huebel, Alish Palmos Johan Zrvskovec, Oliver Pain"
date: "28/08/2021"
output: html_document
---

# Rmd set up

## General set up
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      comment=NA,
                      prompt=FALSE,
                      cache=FALSE)
```

Clear global environment
```{r Clear global environment}
remove(list = ls())
```

Package setup
```{r Packages}
#devtools::install_github("johanzvrskovec/shru")

library(optparse)
library(data.table)
library(shru)
library(googlesheets4)
library(tidyverse)
```

## Command line set up
```{r command line setup}
column_parser <- OptionParser()

column_parser <- add_option(
  object = column_parser,
  opt_str = c("-t", "--task"),
  type = "character",
  default ="clean",
  help = "The explicit task to run:\nclean: Clean GWAS sumstat [default %default]"
  )

column_parser <- add_option(
  object = column_parser,
  opt_str = c("-f", "--file"),
  type = "character",
  help = "The file to process"
  )

column_parser <- add_option(
  object = column_parser,
  opt_str = c("-c", "--code"),
  type = "character",
  help = "The code to use for the dataset from the specified file."
  )

column_parser <- add_option(
  object = column_parser,
  opt_str = c("-n", "--sample-size"),
  type = "numeric",
  help = "Number of individuals in the sample."
  )

column_options <- parse_args(column_parser)
```

Retrieve the recent date
We use this to save files with a time stemp.
```{r Recent date}
date = Sys.Date()
date
```


## Source credentials files
This file includes the file path to the King's College London Microsoft Teams channel to store the raw data files of the GWAS summary statistics (intermediate solution before we move the final version to the rosalind cluster). However, we probably will save all summary statistics raw files on SharePoint/Teams to save space on rosalind.

sumstats_path

The file also contains the login details for your google work account: google_account_details

```{r Credentials file}
source(file = "../credentials/credentials.R")
```


## GWAS summary statistics overview sheet

Login to googlesheets
```{r Login to google sheets}
gs4_auth(email = google_account_details)
```


Import google spreadsheet that contains the information of the GWAS
```{r Google spreadsheet of GWAS dictionary}
gwas_sheet <- read_sheet(
  ss = "https://docs.google.com/spreadsheets/d/1Wmj4Qzumkzks6541vQJqxaUBxjfL4KLQo7_K_oquacI/edit#gid=0",
  sheet = "SGDP_GWASLIST_EDITTHIS",
  col_names = TRUE,
  trim_ws = TRUE, # Trims whitespace,
  guess_max = 600,
  na = c("", "NULL", "NA")
  )

colnames(gwas_sheet)

gwas_sheet
```

## Script settings

### Paths
A list of file paths you want to be processed in parallel. In the function this will be a list, but for the command line utility we have to parse the argument, maybe have paths separated by comma or similar?
_Set your file path_, when testing
```{r Paths}
filePaths <- c(
  paste0(sumstats_path, "/data_raw/CUD_EUR_full_public_11.14.2020.gz")
  )
# filePaths <- column_options$file
filePaths
```

### Trait names
A list of nice trait names which could be fit for a publication table/graph/plot
_Set your nice trait name_, when testing
```{r Trait names}
traitNames <- c("Cannabis use disorder (2020)")
#traitNames <- column_options$names
traitNames
```

### Code
A list of codes you want to identify your traits by (different from the nice names), used to reference a trait by a string
These codes consist of 4 letters; for examples investigate the GWAS spreadsheet. Try to identify similar traits.
_Set your trait code string_, when testing
```{r Code}
traitCodes <- c("CUDI01")
#traitCodes <- column_options$code
traitCodes
```

### Creating the sumstats_meta object
sumstats_meta is an object containing an overview with information of the cleaning steps that is generated throughout the cleaning pipeline for each GWAS trait.
```{r Creating the sumstats_meta object}
sumstats_meta <- data.frame(
  name = traitNames
  )

row.names(sumstats_meta) <- traitCodes
sumstats_meta
```

### Sample size
Explicitly set total N (e.g., a vector of sample sizes)
```{r Sample size}
cN <- column_options$`sample-size`
```

### Indels
Setting for keeping or removing indels
```{r Indels}
keepIndel <- FALSE # or change to TRUE
```

### MHC filter
Activate/deactivate MHC filter
+++CH: What does 37 mean here?
```{r MHC filter}
mhc.filter <- 37
```

### Supermunge
Activate/deactivate supermunge
+++CH: This needs one line explanation what supermunge does.
```{r Supermunge setting}
doSupermunge <- TRUE
```


### Standardised colum names
This chunk reads in a function called "standardise_GWAS_column_names". The function interprets the columns of the original raw GWAS summary statistics and replaces them with standard column names. The standard column names are the first item in the vector:
SNP	CHR	ORIGBP	A1	A2	FREQ	FREQ_CASES	P	INFO	OR/BETA/Z	SE	N	Ncas	Ncon
We have chosen ORIGBP as output name, because the summary statistics are based on different genome builds which means SNPs will be located at different BPs positions.
To avoid follow-up software using the BP column (cave: different genome builds), we named it ORIGBP.
The function stops if the essential columns SNP, A1, A2, P, BETA or OR or Z are not found.

```{r Standardise GWAS column names}
source(file = "../functions/standardise_GWAS_column_names.R")
```

# Chromosome numbering
We are changing naming of chromosomes to PLINK standard.
As a reference, we are using the PLINK naming:
plink chromosome numbering: https://zzz.bwh.harvard.edu/plink/data.shtml

This chunk creates the functions "parse_SNP_column_as_rs_number" and "parse_CHR_column" which parses the SNP and CHR columns respectively. It means that it can read different formats of data in those and convert it to rs-id and chromosome number which we expect to be in the SNP and CHR columns.
```{r Chromosome numbering}
source(file = "../functions/parse_SNP_column_as_rs_number.R")

```



## Read in sumstats and interpret columns
We use the data.table package, because it is much faster than the basic R functions to read in data.
https://cran.r-project.org/web/packages/data.table/vignettes/datatable-intro.html

_Future_
+++JZ: TODO replace commas with field separators as in Helena's script
+++CH: Has this happened?
```{r read in sumstats}
cCode <- traitCodes[1]

cFilePath <- filePaths[1] #we will probably want to have a loop over the provided files

cSumstats <- data.table::fread(
  file = cFilePath,
  header = T,
  quote = "\"",
  fill = T, # If TRUE then in case the rows have unequal length, blank fields are implicitly filled.
  na.strings = c(
    ".",
    NA,
    "NA",
    ""
    )
  )

#number of rows before touching the data
nRowsRaw <- nrow(cSumstats)
sumstats_meta[cCode,c("n_snp_raw")]<-nRowsRaw #add number of SNPs raw to sumstats_meta (meta data file)

sumstats_meta$n_snp_raw
```



## Interpret columns
Checks the existence of columns based on list of column names in chunk above (Standardise GWAS column names). Renames columns to standardised names.
```{r interpret columns}
newNames <- standardise_GWAS_column_names(
  column_names = colnames(cSumstats)
  )
print(newNames)
colnames(cSumstats) <- newNames$std
```

# Column and dataframe re-formatting

## Re-format and parse special format columns
Set correct R data formats

Transform to data table
```{r Transform to data table}
cSumstats <- setDT(cSumstats)
```

Clean or fix variant SNP column
```{r Fix SNP column}
cSumstats.keys <- c('SNP')
cSumstats$SNP <- as.character(cSumstats$SNP)
```

Format A1 and A2 to upper case
```{r Format A1 and A2 to upper case}
cSumstats$A1 <- toupper(as.character(cSumstats$A1))
cSumstats$A2 <- toupper(as.character(cSumstats$A2))
```

Parse SNP (if needed)
```{r Parse SNP}
cSumstats$SNP <- tolower(parse_SNP_column_as_rs_number(cSumstats$SNP))

if(any(colnames(cSumstats)=="CHR")) {
  cSumstats$CHR <- toupper(as.character(cSumstats$CHR))
  cSumstats.keys <- c(cSumstats.keys,'CHR')
}

if(any(colnames(cSumstats)=="BP")) {
  cSumstats$BP <- as.integer(cSumstats$BP)
  cSumstats.keys <- c(cSumstats.keys,'BP')
}
```

Set data table index on selected keys
+++CH: Johan, what does this do, please explain
+++JZ: It orders and indexes the data based on the key columns specified in the cols argument. Any data.table matching using the key columns will be performed more quickly when it has been run previously. It may have to be re-run if you change the data.table too much.
```{r Set data table index on selected keys}
setkeyv(cSumstats,cols = cSumstats.keys)
```


Interpret BGENIE SNP format

Check if daner like columns present
ref: https://docs.google.com/document/d/1TWIhr8-qpCXB13WCXcU1_HDio8lC_MeWoAg2jlggrtU/edit
```{r Interpret daner like format}
if(!any(colnames(cSumstats)=="FRQ") & !any(colnames(cSumstats)=="N_CAS") & !any(colnames(cSumstats)=="N_CON")){
  danerNcas<-colnames(cSumstats)[startsWith(colnames(cSumstats),prefix = "FRQ_A_")][1]
  danerNcon<-colnames(cSumstats)[startsWith(colnames(cSumstats),prefix = "FRQ_U_")][1]
  if(!is.na(danerNcas) & !is.na(danerNcon)){
    #add number of cases and number of controls
    cSumstats$N_CAS <- danerNcas
    cSumstats$N_CON <- danerNcon
    #add case and control specific effect allele frequencies
    colnames(cSumstats)[danerNcas] <- "FRQ_CASES"
    colnames(cSumstats)[danerNcon] <- "FRQ_CONTROLS"
  }
  
}
```


Rename duplicate columns
Deal with duplicate columns - use the first occurrence
```{r Rename duplicate columns}
#+++JZ: Moved this part to the column name standardisation script to de-clutter the main script
```


+++JZ: DO WE WANT TO CONVERT CHR TO NUMERIC VALUES?
```{r Convert to numeric columns}

```


# Add missing columns

## Add total N depending on columns and input
1) Calculate N from Ncon and Ncas, if not present
2) Set N from external argument (set at the beginning)

Add columns on sample size, enough for PRS enough for pathway analysis

```{r add missing N columns}
if(any(colnames(cSumstats)=="N_CAS") && any(colnames(cSumstats)=="N_CON")) {
  ### Calculate total N from number of cases and number of controls if they are present. Overwrite any specific total N.
  cSumstats$N <- cSumstats$N_CAS + cSumstats$N_CON
} else if(!is.null(cN)) {
  cSumstats$N[which(is.na(cSumstats$N))]<-cN #default behaviour - add explicit N only to missing values.
} else if(!any(colnames(cSumstats)=="N")) {
  warning("\nNo N column detected!\n")
  sumstats_meta[cCode,c("noN")]<-T
}
```

## Add BETA column if missing
JZ moved this here so we can rely on BETA existing if either BETA or OR are present in the raw data file. We can't know if the effect is a non-OR unless we check though.
+++CH: Johan, what do you mean by "We can't know if the effect is a non-OR unless we check though."?
+++JZ: You can check if the effect is an odds ratio if the (rounded) median is 1. Other comparisons you can do is comparing the Z-values and P if SE is present while using combinations of log(EFFECT/SE) or e^(EFFECT/SE)
if(round(median(cSumstats$EFFECT,na.rm=T)) == 1) #from Genomic SEM code
```{r add BETA column if missing}
# if no BETA column
if(any(colnames(cSumstats) == "BETA") == 0 & any(colnames(cSumstats) == "OR")){
    cSumstats$BETA<-log(cSumstats$OR) #calculate BETA from log of OR
    cat("BETA column inserted based on log OR value")
    sumstats_meta[cCode,c("Effect")]<-"OR"
} else if(any(colnames(cSumstats) == "BETA")) {
    cat("BETA column detected")
    sumstats_meta[cCode,c("Effect")]<-"BETA"
    }
```



## Add missing Z column
We know that BETA exists if any effect provided previously
+++JZ: Need to be inferred from both the P and the sign of any existing effect. If the effect is not known we cannot know the direction of Z either just from P. Can be calculated from EFFECT and SE but the SE may not be reliable as it can be given on the ln scale or not. There are ways we can check this and maybe it should be part of the light cleaning as well.

```{r insert Z column if missing}
# if no Z column
if(any(colnames(cSumstats) == "Z") == 0 & any(colnames(cSumstats)=="BETA")){
  cSumstats$Z <- sign(cSumstats$BETA)*sqrt(qchisq(cSumstats$P,1,lower=F)) #calculate Z value using qchisq
  cat("Z column inserted based on P value and sign(BETA)")
  sumstats_meta[cCode,c("Z")]<-"P,sign(BETA)"
} else if(any(colnames(cSumstats)=="SE") & any(colnames(cSumstats)=="BETA")){
  cSumstats$Z <- cSumstats$BETA/cSumstats$SE
  cat("Z column inserted based on BETA value and SE")
  sumstats_meta[cCode,c("Z")]<-"BETA,SE"
}
```


## Add missing effect standard errors column 
If missing, add standard errors from beta & z (we always have BETA now if any effect provided previously)
```{r insert SE column if missing}
# if no SE column
if(any(colnames(cSumstats) == "SE") == 0){ 
  if(any(colnames(cSumstats) == "BETA") == 1){ #take BETA if BETA column exists
    cSumstats$SE<-abs(cSumstats$BETA/cSumstats$Z) #calculate SE from BETA and Z
  }
  cat("SE column inserted based on BETA/OR and P")
  sumstats_meta[cCode,c("SE")]<-"BETA,Z"
  }
```


## Meta data about sufficient columns for analyses
Add information to metadata if there are enough columns for PRS and pathway analysis
```{r add meta data for PRS & pathway analysis}
sumstats_meta[cCode,c("enough_columns_PRS")] <- #sufficient columns for PRS analysis
  any(colnames(cSumstats)=="SNP") &
  any(colnames(cSumstats)=="P") &
  (any(colnames(cSumstats)=="BETA") |
     any(colnames(cSumstats)=="OR") |
     any(colnames(cSumstats)=="Z") ) &
  any(colnames(cSumstats)=="A1") &
  any(colnames(cSumstats)=="A2")

sumstats_meta[cCode,c("enough_columns_pathway_analysis")] <- #sufficient columns  for pathway analysis
  any(colnames(cSumstats)=="SNP") &
  any(colnames(cSumstats)=="P")

sumstats_meta
```


# Cleaning

##Missing data

Remove variants with missing P values and missing effect columns
Clean if missing P column
```{r clean missing p}
if(any(colnames(cSumstats)=="P")){
  cond <- is.na(cSumstats$P)
  sumstats_meta[cCode,c("n_removed_missing_p")]<-sum(cond)
  cSumstats<-cSumstats[which(!cond),]
}

sumstats_meta$n_removed_missing_p
```

Clean if missing effect columns 
```{r clean missing effect}
if(any(colnames(cSumstats)=="BETA")){ 
  cond <- is.na(cSumstats$BETA) #if NA for BETA value remove variant
  sumstats_meta[cCode,c("n_removed_missing_effect")]<-sum(cond)
  cSumstats<-cSumstats[which(!cond),]
} else if(any(colnames(cSumstats)=="OR")) {
  cond <- is.na(cSumstats$OR) #if NA for OR value remove variant
  sumstats_meta[cCode,c("n_removed_missing_effect")]<-sum(cond)
  cSumstats<-cSumstats[which(!cond),]
} else if(any(colnames(cSumstats)=="Z")) {
  cond <- is.na(cSumstats$Z) #if NA for Z value remove variant
  sumstats_meta[cCode,c("n_removed_missing_effect")]<-sum(cond)
  cSumstats<-cSumstats[which(!cond),]
}

sumstats_meta$n_removed_missing_effect
```


## Duplicate data

Deal with duplicate variants
```{r clean duplicate data}
# Remove duplicated variants across SNP, A1 and A2
if(any(colnames(cSumstats)=="SNP") & any(colnames(cSumstats)=="A1") & any(colnames(cSumstats)=="A2")) {
  cSumstats.n <- nrow(cSumstats)
  cSumstats <- unique(cSumstats,by = c("SNP","A1","A2"))
  sumstats_meta[cCode,c("n_removed_duplicates")]<-cSumstats.n-nrow(cSumstats)
}

sumstats_meta$n_removed_duplicates
```


## Value filters

?Remove variants based on MAF/FRQ, INFO, OR, p-value, N (Ollies script) filters etc

# Clean based on values

## P filter
```{r P filter}
if(any(colnames(cSumstats)=="P")){
  rm <- (!is.na(cSumstats$P) & (cSumstats$P>1 | cSumstats$P<0)) #if outside of bounds P < 0 or P > 1
  cSumstats <- cSumstats[!rm, ]
  cat("Removing ", sum(rm), " SNPs with P < 0 or P > 1; ", nrow(cSumstats), " remain") 
  sumstats_meta[cCode,c("n_removed_p")]<-sum(rm)
} else {
  cat("Warning: The dataset does not contain a P column to apply the specified filter on.")
}

sumstats_meta$n_removed_p
```

## FRQ filter
```{r FRQ filter}
if(any(colnames(cSumstats)=="FRQ")){
  rm <- (!is.na(cSumstats$FRQ) & (cSumstats$FRQ<0.005 | cSumstats$FRQ>0.995)) #if outside of bounds 
  cSumstats <- cSumstats[!rm, ]
  cat("Removing ", sum(rm), " SNPs with FRQ <", frq.filter, "; ", nrow(cSumstats), " remain")
  sumstats_meta[cCode,c("n_removed_frq")]<-sum(rm)
} else {
  cat("Warning: The dataset does not contain a FRQ or MAF column to apply the specified filter on.")
}

sumstats_meta$n_removed_frq
```


## INFO filter
```{r INFO filter}
if(any(colnames(cSumstats)=="INFO")){
  rm <- (!is.na(cSumstats$INFO) & cSumstats$INFO<0.6)
  cSumstats <- cSumstats[!rm, ]
  cat("Removing ", sum(rm), " SNPs with INFO < 0.6; ", nrow(cSumstats), " remain")
  sumstats_meta[cCode,c("n_removed_info")]<-sum(rm)
} else {
  cat("Warning: The dataset does not contain an INFO column to apply the specified filter on.")
}

sumstats_meta$n_removed_info
```

## OR filter
+++JZ: TODO - adapt to all kinds of effect types
```{r OR filter}
if(any(colnames(cSumstats)=="OR")){
  rm <- (!is.na(cSumstats$OR) & cSumstats$OR>10000)
  cSumstats <- cSumstats[!rm, ]
  cat("Removing ", sum(rm), " SNPs with OR > 10000; ", nrow(cSumstats), " remain")
  sumstats_meta[cCode,c("n_removed_or")]<-sum(rm)
} else {
  cat("Warning: The dataset does not contain an INFO column to apply the specified filter on.")
}

sumstats_meta$n_removed_or
```

## N filter - (from Ollies script)
+++ATK: do we want to remove SNPs that have a large N as set here? Or just SNPs with small N? 
```{r N filter}
if(any(colnames(cSumstats)=="N")){
  N_sd<-sd(cSumstats$N)
  N_median<-median(cSumstats$N)
  rm <- (!is.na(cSumstats$N) & (cSumstats$N > N_median+(3*N_sd) | cSumstats$N < N_median-(3*N_sd)))
  cSumstats <- cSumstats[!rm, ]
  cat("Removing ", sum(rm), " SNPs with N outside median(N) +- 3SD(N); ", nrow(cSumstats), " remain")
  sumstats_meta[cCode,c("n_removed_n")]<-sum(rm)
}

sumstats_meta$n_removed_n
```

# SE filter - from Ollies script
Note: if SE is 0 then this is flagged because it can cause issues in downstream analyses
```{r SE filter}
if(any(colnames(cSumstats)=="SE")){
  rm <- (!is.na(cSumstats$SE) & cSumstats$SE == 0) #if SE is 0 or NA - remove
  cSumstats <- cSumstats[!rm, ]
  cat("Removing ", sum(rm), " SNPs with SE = 0; ", nrow(cSumstats), " remain")
  sumstats_meta[cCode,c("n_removed_se")]<-sum(rm)
}

sumstats_meta$n_removed_se
```

#indels

+++Ollie: Preferable to not remove these as standard in light cleaning.

```{r indels filter}
if(keepIndel == T){ #if set at True, keeps indels
  cSumstats$A1 <- as.character(toupper(cSumstats$A1))
  cSumstats$A2 <- as.character(toupper(cSumstats$A2))
} else if(keepIndel == F){ #if set at False, removes indels
  cSumstats$A1 <- as.character(toupper(cSumstats$A1), c("A", "C", "G", "T"))
  cSumstats$A2 <- as.character(toupper(cSumstats$A2), c("A", "C", "G", "T"))
  rm <- (is.na(cSumstats$A1) | is.na(cSumstats$A2))
  sumstats_meta[cCode,c("n_removed_indels")]<-sum(cond)
  cSumstats<-cSumstats[!rm,]
}

sumstats_meta$n_removed_indels
```


#Check and correct for Genomic Control

+++JZ Only relies on BETA as we know it exists if any effect provided previously

Added from Ollie's script:
```{r check and correct for genomic control}
if(any(colnames(cSumstats) == "SE") == 1){
  if(any(colnames(cSumstats) == "BETA") == 1){
    cSumstats$Z<-cSumstats$BETA/cSumstats$SE ####change to Z_check?
    cSumstats$P_check<-2*pnorm(-abs(cSumstats$Z))
    cSumstats$Z<-NULL

    if(abs(mean(cSumstats$P[!is.na(cSumstats$P_check)]) - mean(cSumstats$P_check[!is.na(cSumstats$P_check)])) > 0.01){
      cSumstats$P<-cSumstats$P_check
      cSumstats$P_check<-NULL
      
      cat("Genomic control detected. P-value recomputed using BETA and SE.")
      sumstats_meta[cCode,c("GC")]<-T
     
    } else {
     
      cat("Genomic control was not detected.")
      sumstats_meta[cCode,c("GC")]<-F
    
      cSumstats$P_check<-NULL
    }
    
  }
} else {
  
  cat("SE column is not present, genomic control cannot be detected.")
 
}


```
 
Need to add from Helenas (below):
Helena: /mnt/lustre/groups/ukbiobank/sumstats/scripts/computeLambdaAndCorrect.r
```{r genomic control Helena's script}
##++++ATK: is there anything we want to add from Helena's below? http://genometoolbox.blogspot.com/2014/08/how-to-calculate-genomic-inflation.html
#arg <- commandArgs(TRUE)
#output <- arg[2]
#
#
#gres <- read.table(arg[1],header=T,sep="\t")
#
#chisq <- qchisq(1-gres$P,1) #check chisq
#
#lambda<-median(chisq)/qchisq(0.5,1) #check lambda
#
#gres$P <- pnorm(-abs((qnorm(gres$P/2)/sqrt(lambda))))*2 #check p 
#
#gres[,"N"]<-floor(gres[,"N"])
#
#gz1 <- gzfile(paste(output,".gz",sep=""),"w")
#write.table(gres,gz1,quote=F,row.names=F,col.names=T,sep="\t")
#close(gz1)
```

# GC calculation: Lambda
The block below relies on the Z value being up to date.
```{r Calculate lambda GC}
#http://genometoolbox.blogspot.com/2014/08/how-to-calculate-genomic-inflation.html

sumstats_meta[cCode,c("lambdaGC")] <- median(cSumstats$Z^2)/qchisq(0.5,1)

cat("\nThe genomic inflation factor (lambda GC) was calculated as: ",round(sumstats_meta[cCode,c("lambdaGC")],digits = 2))

```

#

#Deal with maf and frequencey columns
  

NOTE: This would only work if in the harmonization step earlier, the MAF_REF column was insterted into the cSumstats dataframe (so need to add harmonisation step before this step)

+++Alish: this is slightly different to what we were previously doing in Helena's script. Discuss in meeting?

#Compares to reference and inverts MAFs if they are incorrect based on average allele frequency.
```{r allele Frequency columns and MAF inversion}
# Check FRQ column is valid

if(any(cSumstats$FRQ>1) || any(cSumstats$FRQ<0)) {
      
      stop(paste0('\nThere are FRQ values larger than 1 (',sum(cSumstats$FRQ>1),') or less than 0 (',sum(cSumstats$FRQ<0),') which is outside of the possible FRQ range.')
      )
    }

### Invert FRQ based on the previous reference matching
  if(!is.null(cond.invertedAlleleOrder) & !harmoniseAllelesToReference) {
    
    alleleFRQ <- ifelse(cond.invertedAlleleOrder, (1-cSumstats$FRQ), cSumstats$FRQ)
    
    if(mean(alleleFRQ)<mean(cSumstats$FRQ)){
      
      cSumstats$FRQ<-alleleFRQ
      
      csumstats_meta<-rbind(csumstats_meta,list("FRQ","Fitted (flipped) according to reference allele order"))
      
      sumstats_meta[iFile,c("FRQ.flipped")] <- T
      
      } else {
        
        cond.invertedMAF<-cSumstats$FRQ > .5
        
        cSumstats$FRQ<-ifelse(cond.invertedMAF,1-cSumstats$FRQ,cSumstats$FRQ)
        
        csumstats_meta<-rbind(csumstats_meta,list("FRQ","Set to MAF"))
        
        sumstats_meta[iFile,c("FRQ.flipped")]<-F
        
      }
  }
```

#if allele frequency column is missing, infer from reference panel 
```{r Compute MAF columns based on reference MAF}
### Compute MAF 

  if(cond.invertedMAF<-cSumstats$FRQ > .5){
    
    cSumstats$MAF<-ifelse(cond.invertedMAF,1-cSumstats$FRQ,cSumstats$FRQ)
  
  ### Does not have FRQ
  #### Note that FRQ is not present
  
  if(sumstats_meta[iFile,c("no_FRQ")]<-T){
    
  cSumstats.warnings<-c(cSumstats.warnings,"No FRQ column present!")
  
  }
  
  if(!is.null(ref)){
    
    #Set FRQ from ref if not present
    
    cSumstats$FRQ <- cSumstats$MAF_REF #creates new column with allele frequency inffered from reference
    
    cSumstats.warnings <- c(cSumstats.warnings,"Inferring FRQ from reference!")
    
  }
    
  else {
      
    #### Add empty FRQ here for consistency (sets to NAs)
    
      cSumstats$FRQ<-NA_real_
      
  }
  }

```


# Save cleaned file and log output

+++JZ: When converted to a function we would want to provide this as an argument (with the possibility of a default value).
```{r save cleaned file as .gz file in cleaned folder}
cSumstats %>%
  gzfile(
    file = paste0("../cleaned/", cCode, ".gz", sep="\t"))

```

# Update overview spreadsheet

+++JZ: TODO
```{r update overview metadata spreadsheet}



```


+++CH: Additionally, a log file should be created and be saved in the log file folder
```{r create log file and save in log file folder}

```

?Read in reference from plink formats
```{r read in reference}

```


## MHC removal (separate from cleaning script)
This should happen after the LDSC munging; some people will want to use the MHC variants for their analyses

remove MHC region based on position - probably better to do after merge/harmonisation with reference
references:
https://www.ncbi.nlm.nih.gov/grc/human/regions/MHC?asm=GRCh37
https://www.ncbi.nlm.nih.gov/grc/human/regions/MHC
```{r MHC removal}
if(!is.na(mhc.filter)){
  if(any(colnames(cSumstats)=="CHR") & any(colnames(cSumstats)=="BP")){
    if(mhc.filter==37) {
        rm <- (!is.na(cSumstats$CHR) & !is.na(cSumstats$BP) & cSumstats$CHR=="6" & cSumstats$BP>=28477797 & cSumstats$BP<=33448354)
        cSumstats <- cSumstats[!rm, ]
    } else if (mhc.filter==38) {
        rm <- (!is.na(cSumstats$CHR) & !is.na(cSumstats$BP) & cSumstats$CHR=="6" & cSumstats$BP>=28510120 & cSumstats$BP<=33480577)
        cSumstats <- cSumstats[!rm, ]
        cat("Removing ", sum(rm), " SNPs in the GRCh",mhc.filter," MHC; ", nrow(cSumstats), " remain")
    } else {
        cat("Warning: Invalid assembly version provided - no filtering of the MHC was done!")
      }
  } else {
    cat("Warning: No chromosome or base-pair position information available - no filtering of the MHC was done!")
  }
}
```

#Munge sumstats
```{r}

#Write the trait code into a text file 
c(traitCodes) %>% write_lines( "GWAS_name.txt")

```

```{bash}
#!/bin/bash
#SBATCH -n 8
#SBATCH --mem-per-cpu=9G
#SBATCH -t 72:00:00

###
module add devtools/anaconda/2019.3-python2.7.16

#parameters
input_path="/scratch/groups/ukbiobank/sumstats/cleaned/"
file=`cat GWAS_name.txt`
output_path="/scratch/groups/ukbiobank/sumstats/munged/"
ldscdir="/scratch/groups/ukbiobank/Edinburgh_Data/Software/ldscore/"
#optional parameters to compute h2 liability scale
#Nca=$3
#Nco=$4

#####munge with w_hm3
python ${ldscdir}/munge_sumstats.py --out ${output_path}/${file} --merge-alleles ${ldscdir}/w_hm3.snplist --N-cas-col Ncas --N-con-col Ncon --N-col N --sumstats /scratch/groups/ukbiobank/sumstats/cleaned/${input_path}/${file}.gz --info-min 0.6

```


#Supermunge
```{r Supermunge}
if(doSupermunge){
  supermungeTraitCode<-paste0(traitCodes[1],"_supermunge")
  supermunge(
    list_df = list(cSumstats),
    refFilePath = "/scratch/groups/ukbiobank/Edinburgh_Data/Software/ldscore/w_hm3.snplist",
    traitNames = paste0(traitCodes[1],"_supermunge"),
    pathDirOutput = "/scratch/groups/ukbiobank/sumstats/munged/",
    N = cN,
    keepIndel = keepIndel
      )
  
  Sys.setenv(supermunge_trait_code=supermungeTraitCode)
}
```


#LD score
```{bash}
#!/bin/bash
#SBATCH -n 8
#SBATCH --mem-per-cpu=9G
#SBATCH -t 72:00:00

###
module add devtools/anaconda/2019.3-python2.7.16

#parameters
input_path="/scratch/groups/ukbiobank/sumstats/cleaned/"
file=`cat GWAS_name.txt`
output_path="/scratch/groups/ukbiobank/sumstats/munged/"
ldscdir="/scratch/groups/ukbiobank/KCL_Data/Software/ldsc/"

#run ldsc 
python ${ldscdir}/ldsc.py --h2 ${output_path}/${file}.gz --ref-ld-chr ${ldscdir}/eur_w_ld_chr/ --w-ld-chr ${ldscdir}/eur_w_ld_chr/ --out ${output_path}/${file}_herit

```

#LD score 2
```{bash}
#!/bin/bash
#SBATCH -n 8
#SBATCH --mem-per-cpu=9G
#SBATCH -t 72:00:00
if [[ -v supermunge_trait_code ]]; then
###
module add devtools/anaconda/2019.3-python2.7.16

#parameters
input_path="/scratch/groups/ukbiobank/sumstats/cleaned/"
output_path="/scratch/groups/ukbiobank/sumstats/munged/"
ldscdir="/scratch/groups/ukbiobank/KCL_Data/Software/ldsc/"

#run ldsc 
python ${ldscdir}/ldsc.py --h2 ${output_path}/${supermunge_trait_code}.gz --ref-ld-chr ${ldscdir}/eur_w_ld_chr/ --w-ld-chr ${ldscdir}/eur_w_ld_chr/ --out ${output_path}/${supermunge_trait_code}_herit

fi

```



