---
title: "rg results summary tables template"
author: "Abigail ter Kuile"
date: "28/08/2021"
output: html_document
---

# Rmd set up
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      comment=NA,
                      prompt=FALSE,
                      cache=FALSE)
```

Clear global environment
```{r Clear global environment}
remove(list = ls())
```

Package setup
```{r Packages}
library(optparse)
library(data.table)
library(summarytools)
library(googlesheets4)
library(tidyverse)
```

## Command line set up
```{r command line setup}
clParser <- OptionParser()

clParser <- add_option(
  object = clParser,
  opt_str = c("-t", "--task"),
  type= "character",
  default="clean",
  help = "The explicit task to run:\nclean: Clean GWAS sumstat [default %default]"
  )

clParser <- add_option(
  object = clParser,
  opt_str = c("-f", "--file"),
  type = "character",
  help="The file to process"
  )

clParser <- add_option(
  object = clParser,
  opt_str = c("-c", "--code"),
  type = "character",
  help = "The code to use for the dataset from the specified file."
  )

clParser <- add_option(
  object = clParser,
  opt_str = c("-n", "--sample-size"),
  type = "numeric",
  help = "Number of individuals in the sample."
  )

clOptions <- parse_args(clParser)
```

Retrieve the recent date
We use this to save files with a time stemp.
```{r Recent date}
date = Sys.Date()
date
```

Specify file name
This file name is needed to export the whole file as an R file
```{r Specify file name}
filename.rmd <- "sumstats_cleaning.Rmd"
```

## Source credentials files
This file includes the file path to the King's College London Microsoft Teams channel to store the raw data files of the GWAS summary statistics (intermediate solution before we move the final version to the rosalind cluster). However, we probably will save all summary statistics raw files on SharePoint/Teams to save space on rosalind.
sumstats_path

The file also contains the login details for your google work account: google_account_details

```{r Credentials file}
source(file = "../credentials/credentials.R")
```


# GWAS summary statistics overview sheet

Login to googlesheets
```{r Login to google sheets}
gs4_auth(email = google_account_details)
```


Import google spreadsheet that contains the information of the GWAS
```{r Google spreadsheet of GWAS dictionary}
gwas_sheet <- read_sheet(
  ss = "https://docs.google.com/spreadsheets/d/1Wmj4Qzumkzks6541vQJqxaUBxjfL4KLQo7_K_oquacI/edit#gid=0",
  sheet = "SGDP_GWASLIST_EDITTHIS",
  col_names = TRUE,
  trim_ws = TRUE, # Trims whitespace,
  guess_max = 600,
  na = c("", "NULL", "NA")
  )

colnames(gwas_sheet)

gwas_sheet
```

## Script settings

### Paths
A list of file paths you want to be processed in parallel. In the function this will be a list, but for the command line utility we have to parse the argument, maybe have paths separated by comma or similar?
_Set your file path_, when testing
```{r Paths}
filePaths <- c(
  paste0(sumstats_path, "/data_raw/CUD_EUR_full_public_11.14.2020.gz")
  ) 
# filePaths <- clOptions$file
```

### Trait names
A list of nice trait names which could be fit for a publication table/graph/plot
_Set your nice trait name_, when testing
```{r Trait names}
traitNames <- c("Cannabis use disorder (2020)") 
#traitNames <- clOptions$names
```

### Code
A list of codes you want to identify your traits by (different from the nice names), used to reference a trait by a string
These codes consist of 4 letters; for examples investigate the GWAS spreadsheet. Try to identify similar traits.
 #Set your trait code string_, when testing
```{r Code}
traitCodes <- c("CUDI01")
#traitCodes <- clOptions$code
```

### Creating the sumstats.meta object
```{r Creating the sumstats.meta object}
sumstats.meta <- data.frame(
  name = traitNames
  )

row.names(sumstats.meta) <- traitCodes
```

### Sample size
Explicitly set total N (e.g., a vector of sample sizes)
```{r Sample size}
cN <- clOptions$`sample-size`
```

### Indels
Setting for keeping or removing indels
```{r Sample size}
keepIndel <- FALSE # or change to TRUE
```

### MHC filter
Activate/deactivate MHC filter
```{r Sample size}
mhc.filter <- 37
```

### R script export
Running this will produce an R script with the same name as the Rmd file.
+++CH: Do we really need this in this script or should this be a separate file?
```{r Purl export, include=FALSE, purl=FALSE, eval=FALSE}
knitr::purl(filename.rmd)
```

+++CH: Maybe we want to put this functions in separate R scripts and source them like we do in the other scripts?

### Standardised colum names
This chunk creates one function called "standardise_GWAS_column_names". The function interprets the columns of the original raw GWAS summary statistics and replaces them with standard column names. The standard column names are the first item in the vector:
SNP	CHR	ORIGBP	A1	A2	FREQ	FREQ_CASES	P	INFO	OR/BETA/Z	SE	N	Ncas	Ncon
We have chosen ORIGBP as output name, because the summary statistics are based on different genome builds which means SNPs will be located at different BPs
To avoid follow-up software using the BP column (cave: different genome builds), we named it ORIGBP.
The function stops if the essential columns SNP, A1, A2, P, BETA or OR or Z 

_Future_
+++CH: FREQ_CASES is missing when reading in, NEF is missing when reading in
+++CH: Ollie has code that detects the genome build of sumstats; we could add the chunk for lifting over. (not top priority).
+++CH: Johan, can you try to find the most recent version of Helena's script on the virtual machine?
+++CH: The stopping and the warning needs to be double checked. I think at the moment it only stops for SNP, A1 or A2 and nothing else.
```{r Standardise GWAS column names}
standardise_GWAS_column_names <- function(column_names, stop_on_missing_essential_column = T,
     c.SNP = c(
       "SNP",
       "PREDICTOR",
       "SNPID",
       "MARKERNAME",
       "MARKER_NAME",
       "SNPTESTID",
       "ID_DBSNP49",
       "RSID",
       "ID",
       "RS_NUMBER",
       "MARKER",
       "RS",
       "RSNUMBER",
       "RS_NUMBERS",
       "SNP.NAME",
       "SNP ID",
       "SNP_ID",
       "LOCATIONALID",
       "ASSAY_NAME"
       ),
     c.A1 = c(
       "A1",
       "ALLELE1",
       "ALLELE_1",
       "INC_ALLELE",
       "EA",
       "A1_EFFECT",
       "REF",
       "EFFECT_ALLELE",
       "RISK_ALLELE",
       "EFFECTALLELE",
       "EFFECT_ALL",
       "REFERENCE_ALLELE",
       "REF_ALLELE",
       "REFERENCEALLELE",
       "EA",
       "ALLELE_1",
       "INC_ALLELE",
       "ALLELE1",
       "A",
       "A_1",
       "CODED_ALLELE",
       "TESTED_ALLELE"
       ),
     c.A2 = c(
       "A2",
       "ALLELE2",
       "ALLELE_2",
       "OTHER_ALLELE",
       "NON_EFFECT_ALLELE",
       "DEC_ALLELE",
       "OA",
       "NEA",
       "ALT",
       "A2_OTHER",
       "NONREF_ALLELE",
       "NEFFECT_ALLELE",
       "NEFFECTALLELE",
       "NONEFFECT_ALLELE",
       "OTHER_ALL",
       "OTHERALLELE",
       "NONEFFECTALLELE",
       "ALLELE0",
       "ALLELE_0",
       "ALT_ALLELE",
       "A_0",
       "NONCODED_ALLELE"
       ),
     #c.EFFECT = c(
     #"EFFECT",
     #"OR",
     #"B",
     #"BETA",
     #"LOG_ODDS",
     #"EFFECTS",
     #"SIGNED_SUMSTAT",
     #"EST"
     #),
     c.BETA = c(
       "BETA",
       "B",
       "EFFECT_BETA",
       "EFFECT",
       "EFFECTS",
       "SIGNED_SUMSTAT",
       "EST",
       "GWAS_BETA",
       "EFFECT_A1",
       "EFFECTA1",
       "EFFECT_NW"
       ),
     c.OR = c(
       "OR",
       "LOG_ODDS",
       "OR",
       "ODDS-RATIO",
       "ODDS_RATIO",
       "ODDSRATIO",
       "OR(MINALLELE)",
       "OR.LOGISTIC",
       "OR_RAN",
       "OR(A1)"
       ),
     c.SE = c(
       "SE",
       "STDER",
       "STDERR",
       "STD",
       "STANDARD_ERROR",
       "OR_SE",
       "STANDARDERROR",
       "STDERR_NW",
       "META.SE",
       "SE_DGC",
       "SE.2GC"
       ),
     c.Z = c(
       "Z",
       "ZSCORE",
       "Z-SCORE",
       "ZSTAT",
       "ZSTATISTIC",
       "GC_ZSCORE",
       "BETAZSCALE"
       ),
     c.INFO = c(
       "INFO",
       "IMPINFO",
       "IMPQUALITY",
       "INFO.PLINK",
       "INFO_UKBB"
       ),
     c.P = c(
       "P",
       "PVALUE",
       "PVAL",
       "P_VALUE",
       "GC_PVALUE",
       "WALD_P",
       "P.VAL",
       "GWAS_P",
       "P-VALUE",
       "P-VAL",
       "FREQUENTIST_ADD_PVALUE",
       "P.VALUE",
       "P_VAL",
       "SCAN-P",
       "P.LMM",
       "META.PVAL",
       "P_RAN",
       "P.ADD",
       "P_BOLT_LMM"
       ),
     c.N = c(
       "N",
       "WEIGHT",
       "NCOMPLETESAMPLES",
       "TOTALSAMPLESIZE",
       "TOTALN",
       "TOTAL_N",
       "N_COMPLETE_SAMPLES",
       "N_TOTAL",
       "N_SAMPLES",
       "N_ANALYZED",
       "NSAMPLES",
       "SAMPLESIZE",
       "SAMPLE_SIZE",
       "TOTAL_SAMPLE_SIZE",
       "TOTALSAMPLESIZE"
       ),
     c.N_CAS = c(
       "Ncas",
       "N_CAS",
       "NCASE",
       "N_CASE",
       "N_CASES",
       "NCAS",
       "NCA",
       "NCASES",
       "CASES",
       "CASES_N",
       "FRQ_A" # +++CH: Is this correct here?
       ),
     c.N_CON = c(
       "Ncon",
       "N_CON",
       "NCONTROL",
       "N_CONTROL",
       "N_CONTROLS",
       "NCON",
       "NCO",
       "N_CON",
       "NCONTROLS",
       "CONTROLS",
       "CONTROLS_N",
       "FRQ_U" # +++CH: Is this correct here?
       ),
     c.NEF = c(
       "NEF",
       "NEFF",
       "NEFFECTIVE",
       "NE"
       ),
     #include FRQ_A? # +++CH: Who is this question by?
     c.FRQ = c(
       "FREQ",
       "FRQ",
       "MAF",
       "AF",
       "CEUAF",
       "FREQ1",
       "EAF",
       "FREQ1.HAPMAP",
       "FREQALLELE1HAPMAPCEU",
       "FREQ.ALLELE1.HAPMAPCEU",
       "EFFECT_ALLELE_FREQ",
       "FREQ.A1",
       "F_A",
       "F_U",
       "FREQ_A",
       "FREQ_U",
       "MA_FREQ",
       "MAF_NW",
       "FREQ_A1",
       "A1FREQ",
       "CODED_ALLELE_FREQUENCY",
       "FREQ_TESTED_ALLELE_IN_HRS",
       "EAF_HRC"
       ),
     c.CHR = c(
       "CHR",
       "CH",
       "CHROMOSOME",
       "CHROM",
       "CHR_BUILD38",
       "CHR_BUILD37",
       "CHR_BUILD36",
       "CHR_B38",
       "CHR_B37",
       "CHR_B36",
       "CHR_ID",
       "SCAFFOLD",
       "HG19CHR",
       "CHR.HG19",
       "CHR_HG19",
       "HG18CHR",
       "CHR.HG18",
       "CHR_HG18",
       "CHR_BP_HG19B37",
       "HG19CHRC"
       ),
     c.BP = c(
       "ORIGBP",
       "BP",
       "POS",
       "POSITION",
       "LOCATION",
       "PHYSPOS",
       "GENPOS",
       "CHR_POSITION",
       "POS_B38",
       "POS_BUILD38",
       "POS_B37",
       "POS_BUILD37",
       "BP_HG19B37",
       "POS_B36",
       "POS_BUILD36",
       "POS.HG19",
       "POS.HG18",
       "POS_HG19",
       "POS_HG18",
       "BP_HG19",
       "BP_HG18",
       "BP.GRCH38",
       "BP.GRCH37",
       "POSITION(HG19)",
       "POSITION(HG18)",
       "POS(B38)",
       "POS(B37)"
       )
                                       ){

  #column_names<-cSumstats.names #+++CH: Why is this hashed out?
  
  column_names.upper <- toupper(column_names)
  #names(column_names)<-column_names #+++CH: Why is this hashed out?
  column_names.orig <- column_names
  
  #+++CH: What does this bit do?
  column_names[column_names.upper %in% c.SNP] <- c.SNP[1]
  column_names[column_names.upper %in% c.A1] <- c.A1[1]
  column_names[column_names.upper %in% c.A2] <- c.A2[1]
  #column_names[column_names.upper %in% c.EFFECT] <- c.EFFECT[1] #+++CH: Why is this hashed out?
  #if(any(column_names==c.EFFECT[1])) column_names[column_names.upper %in% c.Z] <- c.Z[1] else column_names[column_names.upper %in% c.Z] <- c.EFFECT[1] #+++CH: Why is this hashed out?
  column_names[column_names.upper %in% c.BETA] <- c.BETA[1]
  column_names[column_names.upper %in% c.OR] <- c.OR[1] 
  column_names[column_names.upper %in% c.Z] <- c.Z[1] 
  column_names[column_names.upper %in% c.SE] <- c.SE[1]
  column_names[column_names.upper %in% c.INFO] <- c.INFO[1]
  column_names[column_names.upper %in% c.P] <- c.P[1]
  column_names[column_names.upper %in% c.N] <- c.N[1]
  column_names[column_names.upper %in% c.N_CAS] <- c.N_CAS[1]
  column_names[column_names.upper %in% c.N_CON] <- c.N_CON[1]
  column_names[column_names.upper %in% c.NEF] <- c.NEF[1]
  column_names[column_names.upper %in% c.FRQ] <- c.FRQ[1]
  column_names[column_names.upper %in% c.CHR] <- c.CHR[1]
  column_names[column_names.upper %in% c.BP] <- c.BP[1]
  
  if(stop_on_missing_essential_column){
# Stop if any of these columns are not found
    if(!any(column_names=="SNP")) stop("\nCould not find the 'SNP' column.\n")
    if(!any(column_names=="A1")) stop("\nCould not find the 'A1' column.\n")
    if(!any(column_names=="A2")) stop("\nCould not find the 'A2' column.\n")
    if(!any(column_names=="BETA") & !any(column_names=="OR") & !any(column_names=="Z")) stop("Could not find any effect column.\n")  
  }
  
  if(!any(column_names=="P")) warning("\nCould not find the P-value column. Standard is 'P'.\n")
  if(!any(column_names=="FRQ")) warning("\nCould not find the 'FRQ' column.\n")
  
# Warn if multiple of these columns are found #+++CH: Do we want to add more checks for duplicates?
  if(sum(column_names=="SNP")>1) warning("\nMultiple 'SNP' columns found!\n")
  if(sum(column_names=="P")>1) warning("\nMultiple 'P' columns found!\n")
  if(sum(column_names=="A1")>1) warning("\nMultiple 'A1' columns found!\n")
  if(sum(column_names=="A2")>1) warning("\nMultiple 'A2' columns found!\n")
  if(sum(column_names=="BETA")>1) warning("\nMultiple 'BETA' columns found!\n")
  if(sum(column_names=="OR")>1) warning("\nMultiple 'OR' columns found!\n")
  if(sum(column_names=="Z")>1) warning("\nMultiple 'Z' columns found!\n")
  if(sum(column_names=="SE")>1) warning("\nMultiple 'SE' columns found!\n")
  if(sum(column_names=="FRQ")>1) warning("\nMultiple 'FRQ' columns found!\n")
  if(sum(column_names=="INFO")>1) warning("\nMultiple 'INFO' columns found!\n")
  
  return(
    data.frame(
      std = column_names,
      orig = column_names.orig
      )
    )
}
```

# Chromosome numbering
We are changing naming of chromosomes to PLINK standard.
As a reference, we are using the PLINK naming:
plink chromosome numbering: https://zzz.bwh.harvard.edu/plink/data.shtml

_Future_
+++CH: These functions need a sentence explaining what they are doing. Who is going to do this?
+++CH: What is this comment about the number of SNPs? Can we move this outside of the chunk including explanation?
+++JKZ: It only checks the top 100,000 SNPs in a file. We don't know what happens if it detects less than 100,000
```{r Chromosome numbering}
parseSNPColumnAsRSNumber <- function(text){
  #decide if BGENIE SNP format using top 100,000 SNPs
  #TODO this condition may be improved to not rely on the number of variants being >100,000
  #test
  #text<-files[[i]]$SNP
  if(sum(grepl(pattern = "^\\d+:\\w+_\\w+_\\w+", x= head(x = text, n=100000)))>90000){
    #extract and format rs-no
    indexesLengths<-regexec(pattern = "^\\d+:(\\w+)_\\w+_\\w+", text=text)
    matches<-regmatches(text,indexesLengths)
    return(lapply(X = matches, FUN = function(x)paste0("rs",x[2])))
  }
  
  text<-sub(pattern = "^XY:",replacement = "25:",x = text)
  text<-sub(pattern = "^X:",replacement = "23:",x = text)
  text<-sub(pattern = "^Y:",replacement = "24:",x = text)
  text<-sub(pattern = "^MT:",replacement = "26:",x = text)
  text<-sub(pattern = "^chr",replacement = "",x = text)
  text<-sub(pattern = "_",replacement = ":",x = text)
  
  return(text)
}

parseCHRColumn <- function(text){
  text<-sub(pattern = "^XY",replacement = "25",x = text)
  text<-sub(pattern = "^X",replacement = "23",x = text)
  text<-sub(pattern = "^Y",replacement = "24",x = text)
  text<-sub(pattern = "^MT",replacement = "26",x = text)
  text<-sub(pattern = "^chr",replacement = "",x = text)
  return(text)
}
```



## Read in sumstats and interpret columns
We use the data.table package, because it is much faster than the basic R functions to read in data.
https://cran.r-project.org/web/packages/data.table/vignettes/datatable-intro.html

_Future_
+++JZ: TODO replace commas with field separators as in Helena's script
+++CH: Has this happened?
```{r read in sumstats}
cCode <- traitCodes[1]

cFilePath <- filePaths[1] #we will probably want to have a loop over the provided files

cSumstats <- data.table::fread(
  file = cFilePath,
  header = T,
  quote = "\"",
  fill = T, # If TRUE then in case the rows have unequal length, blank fields are implicitly filled.
  na.strings = c(
    ".",
    NA,
    "NA",
    ""
    )
  )

#number of rows before touching the data
nRowsRaw <- nrow(cSumstats)
sumstats.meta[cCode,c("n_snp_raw")]<-nRowsRaw

sumstats.meta$n_snp_raw
```

## Interpret columns
Check the existence of columns. Rename columns to standardised names.
```{r interpret columns}
newNames <- standardise_GWAS_column_names(
  column_names = colnames(cSumstats)
  )
print(newNames)
colnames(cSumstats) <- newNames$std
```

# Column and dataframe re-formatting

## Re-format and parse special format columns
Set correct R data formats

Transform to data table
```{r Transform to data table}
cSumstats <- setDT(cSumstats)
```

Clean or fix variant SNP column
```{r Fix SNP column}
cSumstats.keys <- c('SNP')
cSumstats$SNP <- as.character(cSumstats$SNP)
```

Format A1 and A2 to upper case
```{r Format A1 and A2 to upper case}
cSumstats$A1 <- toupper(as.character(cSumstats$A1))
cSumstats$A2 <- toupper(as.character(cSumstats$A2))
```

Parse SNP (if needed)
```{r Parse SNP}
cSumstats$SNP <- tolower(parseSNPColumnAsRSNumber(cSumstats$SNP))

if(any(colnames(cSumstats)=="CHR")) {
  cSumstats$CHR <- toupper(as.character(cSumstats$CHR))
  cSumstats.keys <- c(cSumstats.keys,'CHR')
}

if(any(colnames(cSumstats)=="BP")) {
  cSumstats$BP <- as.integer(cSumstats$BP)
  cSumstats.keys <- c(cSumstats.keys,'BP')
}
```

Set data table index on selected keys
+++CH: Johan, what does this do, please explain
```{r Set data table index on selected keys}
setkeyv(cSumstats,cols = cSumstats.keys)
```


Interpret BGENIE SNP format

Check if daner like columns present
ref: https://docs.google.com/document/d/1TWIhr8-qpCXB13WCXcU1_HDio8lC_MeWoAg2jlggrtU/edit
```{r Interpret daner like format}
if(!any(colnames(cSumstats)=="FRQ") & !any(colnames(cSumstats)=="N_CAS") & !any(colnames(cSumstats)=="N_CON")){
  danerNcas<-colnames(cSumstats)[startsWith(colnames(cSumstats),prefix = "FRQ_A_")][1]
  danerNcon<-colnames(cSumstats)[startsWith(colnames(cSumstats),prefix = "FRQ_U_")][1]
  if(!is.na(danerNcas) & !is.na(danerNcon)){
    #add number of cases and number of controls
    cSumstats$N_CAS <- danerNcas
    cSumstats$N_CON <- danerNcon
    #add case and control specific effect allele frequencies
    colnames(cSumstats)[danerNcas] <- "FRQ_CASES"
    colnames(cSumstats)[danerNcon] <- "FRQ_CONTROLS"
  }
  
}
```


Rename duplicate columns
Deal with duplicate columns - use the first occurrence
```{r Rename duplicate columns}
##SNP
iDup<-grep(pattern = "^SNP$",colnames(cSumstats))
if(length(iDup)>1){
  iDup<-iDup[2:length(iDup)]
  colnames(cSumstats)[iDup]<-"XSNP"
}

##BP
if('BP' %in% names(cSumstats)){
  iDup<-grep(pattern = "^BP$",colnames(cSumstats))
  if(length(iDup)>1){
    iDup<-iDup[2:length(iDup)]
    colnames(cSumstats)[iDup]<-"XBP"
  }
}

##FRQ
if('FRQ' %in% names(cSumstats)){
  iDup<-grep(pattern = "^FRQ$",colnames(cSumstats))
  if(length(iDup)>1){
    iDup<-iDup[2:length(iDup)]
    colnames(cSumstats)[iDup]<-"XFRQ"
  }
}
```


+++JZ: DO WE WANT TO CONVERT CHR TO NUMERIC VALUES?
```{r Convert to numeric columns}

```


# Add missing columns

## Add total N depending on columns and input
1) Calcualte N from Ncon and Ncas, if not present
2) Set N from external argument (set at the beginning)

Add columns on sample size, enough for PRS enough for pathway analysis

```{r add missing N columns}
if(any(colnames(cSumstats)=="N_CAS") && any(colnames(cSumstats)=="N_CON")) {
  ### Calculate total N from number of cases and number of controls if they are present. Overwrite any specific total N.
  cSumstats$N <- cSumstats$N_CAS + cSumstats$N_CON
} else if(!is.null(cN)) {
  cSumstats$N[which(is.na(cSumstats$N))]<-cN #default behaviour - add explicit N only to missing values.
} else if(!any(colnames(cSumstats)=="N")) {
  warning("\nNo N column detected!\n")
  sumstats.meta[cCode,c("noN")]<-T
}
```

## Add BETA column if missing
JZ moved this here so we can rely on BETA existing if either BETA or OR are present in the raw data file. We can't know if the effect is a non-OR unless we check though.
+++CH: Johan, what do you mean by "We can't know if the effect is a non-OR unless we check though."?
```{r add BETA column if missing}
# if no BETA column
if(any(colnames(cSumstats) == "BETA") == 0 & any(colnames(cSumstats) == "OR")){
    cSumstats$BETA<-log(cSumstats$OR) #calculate BETA from log of OR
    cat("BETA column inserted based on log OR value")
    sumstats.meta[cCode,c("Effect")]<-"OR"
} else if(any(colnames(cSumstats) == "BETA")) {
    cat("BETA column detected")
    sumstats.meta[cCode,c("Effect")]<-"BETA"
    }
```



## Add missing Z column
We know that BETA exists if any effect provided previously
+++JZ: Need to be inferred from both the P and the sign of any existing effect. If the effect is not known we cannot know the direction of Z either just from P. Can be calculated from EFFECT and SE but the SE may not be reliable as it can be given on the ln scale or not. There are ways we can check this and maybe it should be part of the light cleaning as well.

```{r insert Z column if missing}
# if no Z column
if(any(colnames(cSumstats) == "Z") == 0 & any(colnames(cSumstats)=="BETA")){
  cSumstats$Z <- sign(cSumstats$BETA)*sqrt(qchisq(cSumstats$P,1,lower=F)) #calculate Z value using qchisq
  cat("Z column inserted based on P value and sign(BETA)")
  sumstats.meta[cCode,c("Z")]<-"P,sign(BETA)"
} else if(any(colnames(cSumstats)=="SE") & any(colnames(cSumstats)=="BETA")){
  cSumstats$Z <- cSumstats$BETA/cSumstats$SE
  cat("Z column inserted based on BETA value and SE")
  sumstats.meta[cCode,c("Z")]<-"BETA,SE"
}
```


## Add missing effect standard errors column 
If missing, add standard errors from beta & z (we always have BETA now if any effect provided previously)
```{r insert SE column if missing}
# if no SE column
if(any(colnames(cSumstats) == "SE") == 0){ 
  if(any(colnames(cSumstats) == "BETA") == 1){ #take BETA if BETA column exists
    cSumstats$SE<-abs(cSumstats$BETA/cSumstats$Z) #calculate SE from BETA and Z
  }
  cat("SE column inserted based on BETA/OR and P")
  sumstats.meta[cCode,c("SE")]<-"BETA,Z"
  }
```


## Meta data
Add metadata about enough for PRS and pathway analysis
```{r add meta data}
sumstats.meta[cCode,c("enough_columns_PRS")] <- 
  any(colnames(cSumstats)=="SNP") &
  any(colnames(cSumstats)=="P") &
  (any(colnames(cSumstats)=="BETA") |
     any(colnames(cSumstats)=="OR") |
     any(colnames(cSumstats)=="Z") ) &
  any(colnames(cSumstats)=="A1") &
  any(colnames(cSumstats)=="A2")

sumstats.meta[cCode,c("enough_columns_pathway_analysis")] <- 
  any(colnames(cSumstats)=="SNP") &
  any(colnames(cSumstats)=="P")

sumstats.meta
```


# Cleaning

##Missing data

Remove variants with missing values
Clean missing Px
```{r clean missing p}
if(any(colnames(cSumstats)=="P")){
  cond <- is.na(cSumstats$P)
  sumstats.meta[cCode,c("n_removed_missing_p")]<-sum(cond)
  cSumstats<-cSumstats[which(!cond),]
}

sumstats.meta$n_removed_missing_p
```

Clean missing effect
```{r clean missing effect}
if(any(colnames(cSumstats)=="BETA")){
  cond <- is.na(cSumstats$BETA)
  sumstats.meta[cCode,c("n_removed_missing_effect")]<-sum(cond)
  cSumstats<-cSumstats[which(!cond),]
} else if(any(colnames(cSumstats)=="OR")) {
  cond <- is.na(cSumstats$OR)
  sumstats.meta[cCode,c("n_removed_missing_effect")]<-sum(cond)
  cSumstats<-cSumstats[which(!cond),]
} else if(any(colnames(cSumstats)=="Z")) {
  cond <- is.na(cSumstats$Z)
  sumstats.meta[cCode,c("n_removed_missing_effect")]<-sum(cond)
  cSumstats<-cSumstats[which(!cond),]
}

sumstats.meta$n_removed_missing_effect
```


## Duplicate data

Deal with duplicate variants
```{r clean duplicate data}
# Remove duplicated variants across SNP, A1 and A2
if(any(colnames(cSumstats)=="SNP") & any(colnames(cSumstats)=="A1") & any(colnames(cSumstats)=="A2")) {
  cSumstats.n <- nrow(cSumstats)
  cSumstats <- unique(cSumstats,by = c("SNP","A1","A2"))
  sumstats.meta[cCode,c("n_removed_duplicates")]<-cSumstats.n-nrow(cSumstats)
}

sumstats.meta$n_removed_duplicates
```


## Value filters

?Remove variants based on MAF/FRQ, INFO, OR, p-value, N (Ollies script) filters etc

# Clean based on values

## P filter
```{r P filter}
if(any(colnames(cSumstats)=="P")){
  rm <- (!is.na(cSumstats$P) & (cSumstats$P>1 | cSumstats$P<0))
  cSumstats <- cSumstats[!rm, ]
  cat("Removing ", sum(rm), " SNPs with P < 0 or P > 1; ", nrow(cSumstats), " remain")
  sumstats.meta[cCode,c("n_removed_p")]<-sum(rm)
} else {
  cat("Warning: The dataset does not contain a P column to apply the specified filter on.")
}

sumstats.meta$n_removed_p
```

## FRQ filter
```{r FRQ filter}
if(any(colnames(cSumstats)=="FRQ")){
  rm <- (!is.na(cSumstats$FRQ) & (cSumstats$FRQ<0.005 | cSumstats$FRQ>0.995))
  cSumstats <- cSumstats[!rm, ]
  cat("Removing ", sum(rm), " SNPs with FRQ <", frq.filter, "; ", nrow(cSumstats), " remain")
  sumstats.meta[cCode,c("n_removed_frq")]<-sum(rm)
} else {
  cat("Warning: The dataset does not contain a FRQ or MAF column to apply the specified filter on.")
}

sumstats.meta$n_removed_frq
```


## INFO filter
```{r INFO filter}
if(any(colnames(cSumstats)=="INFO")){
  rm <- (!is.na(cSumstats$INFO) & cSumstats$INFO<0.6)
  cSumstats <- cSumstats[!rm, ]
  cat("Removing ", sum(rm), " SNPs with INFO < 0.6; ", nrow(cSumstats), " remain")
  sumstats.meta[cCode,c("n_removed_info")]<-sum(rm)
} else {
  cat("Warning: The dataset does not contain an INFO column to apply the specified filter on.")
}

sumstats.meta$n_removed_info
```

## OR filter
+++JZ: TODO - adapt to all kinds of effect types
```{r OR filter}
if(any(colnames(cSumstats)=="OR")){
  rm <- (!is.na(cSumstats$OR) & cSumstats$OR>10000)
  cSumstats <- cSumstats[!rm, ]
  cat("Removing ", sum(rm), " SNPs with OR > 10000; ", nrow(cSumstats), " remain")
  sumstats.meta[cCode,c("n_removed_or")]<-sum(rm)
} else {
  cat("Warning: The dataset does not contain an INFO column to apply the specified filter on.")
}

sumstats.meta$n_removed_or
```

## N filter - (from Ollies script)
```{r N filter}
if(any(colnames(cSumstats)=="N")){
  N_sd<-sd(cSumstats$N)
  N_median<-median(cSumstats$N)
  rm <- (!is.na(cSumstats$N) & (cSumstats$N > N_median+(3*N_sd) | cSumstats$N < N_median-(3*N_sd)))
  cSumstats <- cSumstats[!rm, ]
  cat("Removing ", sum(rm), " SNPs with N outside median(N) +- 3SD(N); ", nrow(cSumstats), " remain")
  sumstats.meta[cCode,c("n_removed_n")]<-sum(rm)
}

sumstats.meta$n_removed_n
```

# SE filter - from Ollies script
```{r SE filter}
if(any(colnames(cSumstats)=="SE")){
  rm <- (!is.na(cSumstats$SE) & cSumstats$SE == 0)
  cSumstats <- cSumstats[!rm, ]
  cat("Removing ", sum(rm), " SNPs with SE = 0; ", nrow(cSumstats), " remain")
  sumstats.meta[cCode,c("n_removed_se")]<-sum(rm)
}

sumstats.meta$n_removed_se
```

#indels
+++Ollie: Preferable to not remove these as standard in light cleaning.
```{r indels filter}
if(keepIndel == T){
  cSumstats$A1 <- as.character(toupper(cSumstats$A1))
  cSumstats$A2 <- as.character(toupper(cSumstats$A2))
} else if(keepIndel == F){
  cSumstats$A1 <- as.character(toupper(cSumstats$A1), c("A", "C", "G", "T"))
  cSumstats$A2 <- as.character(toupper(cSumstats$A2), c("A", "C", "G", "T"))
  rm <- (is.na(cSumstats$A1) | is.na(cSumstats$A2))
  sumstats.meta[cCode,c("n_removed_indels")]<-sum(cond)
  cSumstats<-cSumstats[!rm,]
}

sumstats.meta$n_removed_indels
```

## MHC removal
+++CH: This should happen after the LDSC munging; some people will want to use the MHC variants for their analyses

remove MHC region based on position - probably better to do after merge/harmonisation with reference
references:
https://www.ncbi.nlm.nih.gov/grc/human/regions/MHC?asm=GRCh37
https://www.ncbi.nlm.nih.gov/grc/human/regions/MHC
```{r MHC removal}
if(!is.na(mhc.filter)){
  if(any(colnames(cSumstats)=="CHR") & any(colnames(cSumstats)=="BP")){
    if(mhc.filter==37) {
        rm <- (!is.na(cSumstats$CHR) & !is.na(cSumstats$BP) & cSumstats$CHR=="6" & cSumstats$BP>=28477797 & cSumstats$BP<=33448354)
        cSumstats <- cSumstats[!rm, ]
    } else if (mhc.filter==38) {
        rm <- (!is.na(cSumstats$CHR) & !is.na(cSumstats$BP) & cSumstats$CHR=="6" & cSumstats$BP>=28510120 & cSumstats$BP<=33480577)
        cSumstats <- cSumstats[!rm, ]
        cat("Removing ", sum(rm), " SNPs in the GRCh",mhc.filter," MHC; ", nrow(cSumstats), " remain")
    } else {
        cat("Warning: Invalid assembly version provided - no filtering of the MHC was done!")
      }
  } else {
    cat("Warning: No chromosome or base-pair position information available - no filtering of the MHC was done!")
  }
}
```



Deal with maf and frequencey columns
+++CH: Please add this for the next meeting from Helena's script
Helena: /mnt/lustre/groups/ukbiobank/sumstats/scripts/submit_freq_2.sh





#Check and correct for Genomic Control
+++JZ Only relies on BETA as we know it exists if any effect provided previously

Added from Ollie's script:
```{r check and correct for genomic control}
if(any(colnames(cSumstats) == "SE") == 1){
  if(any(colnames(cSumstats) == "BETA") == 1){
    cSumstats$Z<-cSumstats$BETA/cSumstats$SE ####change to Z_check?
    cSumstats$P_check<-2*pnorm(-abs(cSumstats$Z))
    cSumstats$Z<-NULL

    if(abs(mean(cSumstats$P[!is.na(cSumstats$P_check)]) - mean(cSumstats$P_check[!is.na(cSumstats$P_check)])) > 0.01){
      cSumstats$P<-cSumstats$P_check
      cSumstats$P_check<-NULL
      
      cat("Genomic control detected. P-value recomputed using BETA and SE.")
      sumstats.meta[cCode,c("GC")]<-T
     
    } else {
     
      cat("Genomic control was not detected.")
      sumstats.meta[cCode,c("GC")]<-F
    
      cSumstats$P_check<-NULL
    }
    
  }
} else {
  
  cat("SE column is not present, genomic control cannot be detected.")
 
}


```
 
Need to add from Helenas (below):
Helena: /mnt/lustre/groups/ukbiobank/sumstats/scripts/computeLambdaAndCorrect.r
```{r genomic control Helena's script}
##++++ATK: is there anything we want to add from Helena's below? http://genometoolbox.blogspot.com/2014/08/how-to-calculate-genomic-inflation.html
#arg <- commandArgs(TRUE)
#output <- arg[2]
#
#
#gres <- read.table(arg[1],header=T,sep="\t")
#
#chisq <- qchisq(1-gres$P,1) #check chisq
#
#lambda<-median(chisq)/qchisq(0.5,1) #check lambda
#
#gres$P <- pnorm(-abs((qnorm(gres$P/2)/sqrt(lambda))))*2 #check p 
#
#gres[,"N"]<-floor(gres[,"N"])
#
#gz1 <- gzfile(paste(output,".gz",sep=""),"w")
#write.table(gres,gz1,quote=F,row.names=F,col.names=T,sep="\t")
#close(gz1)
```

# Yet another GC calculation
Yes, Helena's calculation seems how it is supposed to be done. The block below relies on the Z value being up to date.
```{r Calculate lambda gc}
#http://genometoolbox.blogspot.com/2014/08/how-to-calculate-genomic-inflation.html
sumstats.meta[cCode,c("lambdaGC")] <- median(cSumstats$Z^2)/qchisq(0.5,1)
cat("\nThe genomic inflation factor (lambda GC) was calculated as: ",round(sumstats.meta[cCode,c("lambdaGC")],digits = 2))

```


# Save cleaned file and log output
+++CH: Here the script should end and a cleaned version of the file should be written into the cleaned folder
+++JZ: When converted to a function we would want to provide this as an argument (with the possibility of a default value).
```{r save cleaned file as .gz file in cleaned folder}
cSumstats %>%
  gzfile(
    file = paste0("../cleaned/", cCode, ".gz", sep="\t"))

```

# Update overview spreadsheet

+++JZ: TODO
```{r update overview metadata spreadsheet}



```


+++CH: Additionally, a log file should be created and be saved in the log file folder
```{r create log file and save in log file folder}

```



?Read in reference from plink formats
```{r read in reference}

```

2) Read in Up to date table with full GWAS information of each trait. Currently in csv format linked with the flask app, or in the google sheet.
```{r read in full GWAS info all traits}

```

#Deal with maf and frequencey columns
NOTE: This would only work if in the harmonization step earlier, the MAF_REF column was insterted into the cSumstats dataframe

```{r}

# Check FRQ column is valid

if(any(cSumstats$FRQ>1) || any(cSumstats$FRQ<0)) {
      
      stop(paste0('\nThere are FRQ values larger than 1 (',sum(cSumstats$FRQ>1),') or less than 0 (',sum(cSumstats$FRQ<0),') which is outside of the possible FRQ range.')
      )
    }

### Invert FRQ based on the previous reference matching
  if(!is.null(cond.invertedAlleleOrder) & !harmoniseAllelesToReference) {
    
    alleleFRQ <- ifelse(cond.invertedAlleleOrder, (1-cSumstats$FRQ), cSumstats$FRQ)
    
    if(mean(alleleFRQ)<mean(cSumstats$FRQ)){
      
      cSumstats$FRQ<-alleleFRQ
      
      cSumstats.meta<-rbind(cSumstats.meta,list("FRQ","Fitted (flipped) according to reference allele order"))
      
      sumstats.meta[iFile,c("FRQ.flipped")] <- T
      
      } else {
        
        cond.invertedMAF<-cSumstats$FRQ > .5
        
        cSumstats$FRQ<-ifelse(cond.invertedMAF,1-cSumstats$FRQ,cSumstats$FRQ)
        
        cSumstats.meta<-rbind(cSumstats.meta,list("FRQ","Set to MAF"))
        
        sumstats.meta[iFile,c("FRQ.flipped")]<-F
        
      }
  }

### Compute MAF

  if(cond.invertedMAF<-cSumstats$FRQ > .5){
    
    cSumstats$MAF<-ifelse(cond.invertedMAF,1-cSumstats$FRQ,cSumstats$FRQ)
  
  ### Does not have FRQ
  #### Note that FRQ is not present
  
  if(sumstats.meta[iFile,c("no_FRQ")]<-T){
    
  cSumstats.warnings<-c(cSumstats.warnings,"No FRQ column present!")
  
  }
  
  if(!is.null(ref)){
    
    #Set FRQ from ref if not present
    
    cSumstats$FRQ <- cSumstats$MAF_REF
    
    cSumstats.warnings <- c(cSumstats.warnings,"Inferring FRQ from reference!")
    
  }
    
  else {
      
    #### Add empty FRQ here for consistency
    
      cSumstats$FRQ<-NA_real_
      
  }
  }

```


Deal with missing effect standard error
Add standard errors from beta & z or or & z (Ollies script)







Check and correct for GC



?Read in reference from plink formats
```{r read in reference}

```

2) Read in Up to date table with full GWAS information of each trait. Currently in csv format linked with the flask app, or in the google sheet.
```{r read in full GWAS info all traits}

```

#Munge sumstats
```{r}

#Write the trait code into a text file 
c(traitCodes) %>% write_lines( "GWAS_name.txt")

```

```{bash}
#!/bin/bash
#SBATCH -n 8
#SBATCH --mem-per-cpu=9G
#SBATCH -t 72:00:00

###
module add devtools/anaconda/2019.3-python2.7.16

#parameters
input_path="/scratch/groups/ukbiobank/sumstats/cleaned/"
file=`cat GWAS_name.txt`
output_path="/scratch/groups/ukbiobank/sumstats/munged/"
ldscdir="/scratch/groups/ukbiobank/Edinburgh_Data/Software/ldscore/"
#optional parameters to compute h2 liability scale
#Nca=$3
#Nco=$4

#####munge with w_hm3
python ${ldscdir}/munge_sumstats.py --out ${output_path}/${file} --merge-alleles ${ldscdir}/w_hm3.snplist --N-cas-col Ncas --N-con-col Ncon --N-col N --sumstats /scratch/groups/ukbiobank/sumstats/cleaned/${input_path}/${file}.gz --info-min 0.6

```

#LD score
```{bash}
#!/bin/bash
#SBATCH -n 8
#SBATCH --mem-per-cpu=9G
#SBATCH -t 72:00:00

###
module add devtools/anaconda/2019.3-python2.7.16

#parameters
input_path="/scratch/groups/ukbiobank/sumstats/cleaned/"
file=`cat GWAS_name.txt`
output_path="/scratch/groups/ukbiobank/sumstats/munged/"
ldscdir="/scratch/groups/ukbiobank/KCL_Data/Software/ldsc/"

#run ldsc 
python ${ldscdir}/ldsc.py --h2 ${output_path}/${file}.sumstats.gz --ref-ld-chr ${ldscdir}/eur_w_ld_chr/ --w-ld-chr ${ldscdir}/eur_w_ld_chr/ --out ${output_path}/${file}_herit

```



